### PR TITLE
Use native UPSERTs where possible

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

**Comment by richvdh** [comment]
apparently this is now ready for review

File: synapse/storage/_base.py [hunk 1]
```diff
@@ -537,12 +603,45 @@ def _simple_upsert(self, table, keyvalues, values,
 
                 # presumably we raced with another transaction: let's retry.
                 logger.warn(
-                    "IntegrityError when upserting into %s; retrying: %s",
-                    table, e
+                    "%s when upserting into %s; retrying: %s", e.__name__, table, e
```
Comment (erikjohnston) [inline synapse/storage/_base.py]: Why this change?

File: synapse/storage/_base.py [hunk 2]
```diff
@@ -569,12 +668,56 @@ def _simple_upsert_txn(self, txn, table, keyvalues, values, insertion_values={},
...
+            # We only care about serialization errors, so check for it
+            if e.args[0] == "could not serialize access due to concurrent update":
+                # A concurrent update problem is when we try and do a native
+                # UPSERT but the row has changed from under us. We can either
+                # retry, or give up if asked to do so.
+                if best_effort:
+                    # If it's a concurrent-update problem, and this is marked as
+                    # 'best effort' (i.e. if there's a race, then the one we
+                    # raced with will suffice), then pretend that we succeeded.
+                    return False
```
Comment (erikjohnston) [inline synapse/storage/_base.py]: It's not safe to ignore database errors as the transaction will no longer be valid. This should probably be raised up into `runInteraction` or the like instead.

File: synapse/storage/engines/postgres.py [hunk 3]
```diff
@@ -38,6 +38,18 @@ def convert_param_style(self, sql):
...
+        # Get the version of PostgreSQL that we're using. As per the psycopg2
+        # docs: The number is formed by converting the major, minor, and
+        # revision numbers into two-decimal-digit numbers and appending them
+        # together. For example, version 8.1.5 will be returned as 80105.
+        server_version = str(db_conn.server_version)
+        self._version = (
+            int(server_version[:-4], 10),
+            int(server_version[-4:-2], 10),
+            int(server_version[-2:], 10)
+        )
```
Comment (erikjohnston) [inline synapse/storage/engines/postgres.py]: This feels fragile. Why not do `server_version >= 90500`? The docs also suggest doing `server_version/100 >= 905` to determine feature compatibility based on major versions.

File: synapse/storage/_base.py [hunk 4]
```diff
@@ -569,12 +668,56 @@ def _simple_upsert_txn(self, txn, table, keyvalues, values, insertion_values={},
...
+            ", ".join(k for k in allvalues),
+            ", ".join("?" for _ in allvalues),
+            ", ".join(k for k in keyvalues),
+            ", ".join(k + "=EXCLUDED." + k for k in values),
+        )
+        try:
+            txn.execute(sql, list(allvalues.values()))
+        except self.database_engine.module.OperationalError as e:
+            # We only care about serialization errors, so check for it
+            if e.args[0] == "could not serialize access due to concurrent update":
```
Comment (erikjohnston) [inline synapse/storage/_base.py]: This is probably not safe, since I would be surprised if psycopg2 guaranteed that the error messages are stable. Please just use the error codes for this, as per: https://github.com/matrix-org/synapse/blob/b60749a1ec22269216a503ccd35aaf4454dc507d/synapse/storage/engines/postgres.py#L57-L63

**Comment by erikjohnston** [comment]
Can you please either a) not include unrelated changes in your PRs or b) have sensible commits describing what each change is doing please? It's quite confusing otherwise. (This doesn't include obvious automatic reformatting in edited files) :slightly_smiling_face:

File: synapse/storage/_base.py (hunk 4 above)
Comment (hawkowl) [reply synapse/storage/_base.py]: I was reading off https://www.postgresql.org/docs/10/transaction-iso.html#XACT-REPEATABLE-READ , that didn't give anything nice. That works better, though.

File: synapse/storage/_base.py (hunk 1 above)
Comment (hawkowl) [reply synapse/storage/_base.py]: Oh, this was from my attempt at making a strategy similar to https://github.com/matrix-org/synapse/pull/4306#issuecomment-450564229 work. It didn't. :(

File: synapse/storage/_base.py [hunk 5]
```diff
@@ -569,12 +668,56 @@ def _simple_upsert_txn(self, txn, table, keyvalues, values, insertion_values={},
...
+        """
+        allvalues = {}
+        allvalues.update(keyvalues)
+        allvalues.update(values)
+        allvalues.update(insertion_values)
+
+        sql = (
+            "INSERT INTO %s (%s) VALUES (%s) "
+            "ON CONFLICT (%s) DO UPDATE SET %s "
+            "RETURNING (xmax = 0) AS inserted"
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/_base.py]: Where is it documented? All I could find was the basic system column documentation for it, which doesn't specify anything about upserts, and it doesn't appear on the insert documentation :(

File: synapse/storage/_base.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/_base.py]: Yup, transactions will be rolled back on error, but they're then effectively closed. Note that the link also says that you need to abort and retry the transaction

File: synapse/storage/_base.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/_base.py]: Ah, fair

File: synapse/storage/engines/__init__.py [hunk 6]
```diff
@@ -18,7 +18,7 @@
 
 from ._base import IncorrectDatabaseSetup
 from .postgres import PostgresEngine
-from .sqlite3 import Sqlite3Engine
```
Comment (erikjohnston) [inline synapse/storage/engines/__init__.py]: Why the rename of the file? It seems odd when we haven't renamed the class.

File: synapse/storage/engines/postgres.py [hunk 7]
```diff
@@ -54,6 +61,13 @@ def on_new_connection(self, db_conn):
 
         cursor.close()
 
+    @property
+    def can_native_upsert(self):
+        """
+        Can we use native UPSERTs? This requires PostgreSQL 9.5+.
+        """
+        return self._version >= 90500
```
Comment (erikjohnston) [inline synapse/storage/engines/postgres.py]: This is going to blow up if called before `on_new_connection`. This is probably unlikely, but it may be worth defaulting it to something like `0`. Or have a `_can_do_native_upsert` parameter set to `False` by default instead, and have `on_new_connection` update it when its called

File: synapse/storage/engines/sqlite.py [hunk 8]
```diff
@@ -30,6 +30,21 @@ def __init__(self, database_module, database_config):
...
+        more work we haven't done yet to tell what was inserted vs updated.
+        """
+        # SQLite currently doesn't have a way of telling if an UPSERT ended up
+        # with an INSERT or an UPDATE
+        # To enable, uncomment:
+        #
+        # from sqlite3 import sqlite_version_info
+        # return sqlite_version_info >= (3, 24, 0)
+
+        return False
```
Comment (erikjohnston) [inline synapse/storage/engines/sqlite.py]: Why not just do `can_native_upsert = False`?

File: tox.ini [hunk 9]
```diff
@@ -149,4 +149,5 @@ deps =
     codecov
 commands =
     coverage combine
+    coverage xml
```
Comment (erikjohnston) [inline tox.ini]: What does this do?

File: .coveragerc [hunk 10]
```diff
@@ -1,11 +1,7 @@
 [run]
 branch = True
 parallel = True
-source = synapse
-
-[paths]
-source=
-   coverage
```
Comment (erikjohnston) [inline .coveragerc]: What does this do?

File: MANIFEST.in [hunk 11]
```diff
@@ -9,7 +9,9 @@ include demo/*.py
 include demo/*.sh
 
 recursive-include synapse/storage/schema *.sql
+recursive-include synapse/storage/schema *.sql.*
 recursive-include synapse/storage/schema *.py
+recursive-include synapse/storage/schema *.md
```
Comment (erikjohnston) [inline MANIFEST.in]: We don't seem to have added any of these sorts of files?

File: synapse/storage/_base.py [hunk 12]
```diff
@@ -192,6 +192,41 @@ def __init__(self, db_conn, hs):
 
         self.database_engine = hs.database_engine
 
+        # A set of tables that are not safe to use native upserts in.
+        self._unsafe_to_upsert_tables = {"user_ips"}
```
Comment (erikjohnston) [inline synapse/storage/_base.py]: Presumably the idea here is that we add to this whenever we add an index that we may want to upsert against? This feels fairly fragile tbh, since it'll probably work in CI and then break in production. I wonder if we can hook into the `register_background_index_update` to add to this list when adding a unique index?

Its possibly fine for now

File: synapse/storage/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/_base.py]: I'm still concerned we're relying on an internal detail of how updates and vacuuming of rows work here

File: synapse/storage/_base.py (hunk 12 above)
Comment (erikjohnston) [reply synapse/storage/_base.py]: I ran this past @richvdh, and broadly the conclusion is that this should be an edge case (since we'll aim to add unique indices from the start to new tables), so we can leave it for now. (https://matrix.to/#/!yZHTGeDKZUeKaqeTeU:matrix.org/$1548086288327WrmxW:jki.re?via=matrix.org&via=half-shot.uk&via=sw1v.org is the conversation)

File: synapse/storage/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/_base.py]: TBH, I wonder if we can just get rid of the places where we care about inserted vs updated?

File: synapse/storage/engines/__init__.py (hunk 6 above)
Comment (hawkowl) [reply synapse/storage/engines/__init__.py]: because it shadows a global package :/

File: tox.ini (hunk 9 above)
Comment (hawkowl) [reply tox.ini]: precomputes the xml for codecov. probably don't need it, was experimenting.

File: MANIFEST.in (hunk 11 above)
Comment (hawkowl) [reply MANIFEST.in]: ah, that was from when I had the full schema (which I pulled out of this PR). Will remove.

File: .coveragerc (hunk 10 above)
Comment (hawkowl) [reply .coveragerc]: fixes our coverage to actually cover the storage APIs :(

File: synapse/storage/engines/__init__.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/storage/engines/__init__.py]: Ooooooh, fair :+1:

File: synapse/storage/_base.py [hunk 13]
```diff
@@ -537,12 +586,48 @@ def _simple_upsert(self, table, keyvalues, values,
...
+        keyvalues,
+        values,
+        insertion_values={},
+        lock=True,
+        best_effort=False,
+    ):
+        """
+        Pick the UPSERT method which works best on the platform. Either the
+        native one (Pg9.5+, recent SQLites), or fall back to an emulated method.
+        """
```
Comment (erikjohnston) [inline synapse/storage/_base.py]: Sowwy, but can you add a docstring for these args please? Probably copying and pasting from the non-txn version.

File: synapse/storage/_base.py [hunk 14]
```diff
@@ -515,6 +558,7 @@ def _simple_upsert(self, table, keyvalues, values,
             insertion_values (dict): additional key/values to use only when
                 inserting
             lock (bool): True to lock the table when doing the upsert.
+            best_effort (bool): If we run into a transaction error, do we stop trying?
```
Comment (erikjohnston) [inline synapse/storage/_base.py]: Can you also document the fact that we return `None` if we use a native upsert?

File: synapse/storage/user_directory.py [hunk 15]
```diff
@@ -185,20 +185,38 @@ def _update_profile_in_user_dir_txn(txn):
...
+                            )
+                        """
+                        txn.execute(
+                            sql,
+                            (
+                                user_id, get_localpart_from_id(user_id),
+                                get_domain_from_id(user_id), display_name,
+                            )
+                        )
+                    else:
```
Comment (erikjohnston) [inline synapse/storage/user_directory.py]: Can instead we do:

```python
# TODO: Remove this code after we've bumped the minimum version of postgres to
# always support upserts, so we can get rid of `new_entry` usage
if new_entry is True:
    ....
elif new_entry is False:
    ....
else:
    raise RuntimeError("upsert returned None when 'can_native_upsert' is False")
```

Just as a paranoia check :)

File: synapse/storage/_base.py (hunk 13 above)
Comment (hawkowl) [reply synapse/storage/_base.py]: done

File: synapse/storage/_base.py (hunk 14 above)
Comment (hawkowl) [reply synapse/storage/_base.py]: done

File: synapse/storage/user_directory.py (hunk 15 above)
Comment (hawkowl) [reply synapse/storage/user_directory.py]: done

File: synapse/storage/engines/postgres.py (hunk 7 above)
Comment (hawkowl) [reply synapse/storage/engines/postgres.py]: This won't ever be called before we're actually connected, since it's in transaction code?

File: synapse/storage/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/storage/_base.py]: done
